### DISCUSSION TITLE
Smart ! Cura 5 sucks without your features!!! Reconsider?

### DISCUSSION THREAD

**Discussion by PhilBaz** [smartavionics/Cura#177] [General]
Well its not that it sucks...but honestly the features you add turn Cura from usable to excellent...

I cant get by without fan settings for overhangs! I just cant.

Going back and forth these days as the variable engine is to good to ignore.....But your branch keeps me coming back!

Just wanted to say your efforts are appreciated!

**Comment by smartavionics** [comment]
Hi, thanks for the kind comments. As to merging my build into Cura 5, I have no plans at this time to do that but I will review that decision after UM have made a few more releases from the Cura 5 line and have fixed the worst of the bugs.

**Reply by PhilBaz** [reply]
Well, Ill cross my fingers and hope its goes one day....

Just wanted to show my appreciation for your efforts with master branch... your features truly make the difference and elevate Cura's potential significantly..

**Reply by d-portero** [reply]
I just want to echo what was said. Your additions to Cura are very much needed and I would love for you to implement them in 5.0

**Reply by dprossner** [reply]
I agree with everyone else. the Arachne engine is a game changer and I would love to use it with that bit of extra magic you add.

**Reply by dprossner** [reply]
> Hi, thanks for the kind comments. As to merging my build into Cura 5, I have no plans at this time to do that but I will review that decision after UM have made a few more releases from the Cura 5 line and have fixed the worst of the bugs.

It looks like Ultimaker made some changes to make building Cura 5 easier including Mac and Linux. Changes were mentioned in the 'whats new section' of 5.2 and linked to [here ](https://github.com/Ultimaker/Cura/wiki/Running-Cura-from-Source)

any chance we'll be seeing your optimizations in Cura 5.2?

**Reply by smartavionics** [reply]
I will not be merging my work into Cura 5 but, hey, it's open source, so if anybody wants to take what I've done and port it to Cura 5, they are very welcome to do so.

**Reply by shyblower** [reply]
Fan speed override for overhangs is one of the things I miss most, when using Cura 5 without your additions. I'm really considering to fork 5 and at least implement this feature.

**Reply by shyblower** [reply]
Today I've ported your "Overhanging Wall Fan Speed" and also the "Cooling Fan Lag", because without that the former wouldn't be quite as useful,  features over to Cura 5.2.1.
I had to strip it down a bit because I didn't port your fancy overhanging wall and bridge detection but used the stock one, but so far it seems to work correctly.
The produced G-Code looks good and also the visual and acoustic impression, when examining the fan during printing, tells me that it works as expected.

Are you ok with me posting this publicly as a Cura 5 fork on Github?
I would of course mention your project and the parts I ported from it.
I would also like to use this as a starting point for further Cura 5 modding.

I'll still have to figure how to build an AppImage and maybe Windows and MacOS releases.

**Reply by Sophist-UK** [reply]
@shyblower My opinion as a lurker (but experienced Open Source contributor) is as follows:

a. You are of course entitled to reap the kudos benefits of your own work if you want to e.g. by creating your own fork of Cura with @smartavionics improvements added. But if someone else adds a different set of Mark's enhancements to their own fork, things might start to get messy.

b. The alternative is for Mark to create a v5 branch in his repo that you can PR your changes to - that way this repo remains as **the** fork to come to for enhancements.

**Reply by shyblower** [reply]
@Sophist-UK I would of course also be ok with that as it e.g. would spare me the maintenance burden ;)
... but I would still suggest to give the enhanced Cura fork its own name (like what SuperSlicer is for PrusaSlicer) and a clearly visible description of all the fixes and additional features so that it becomes more visible and accessible to the public.

**Reply by Sophist-UK** [reply]
@shyblower 

Maintenance burden - that too. But I am more concerned about having a single point for community enhancements so as not to create fragmentation.

Fork name - I think in the end, whether you rename depends on whether you are trying to offer community enhancements to the Cura version or to compete as a separate brand with Cura. If you create your own fork, that choice is yours. If you use @smartavionics fork, then that choice is Mark's.

**Reply by smartavionics** [reply]
Hi @shyblower , yes, it's fine, use any of my code anyway you wish. An attribution would be nice but at the end of the day, I don't care that much any more.

I have no plans to merge my code with Cura 5 so it would be better if you make your own fork from the UM repo. Good luck!

**Reply by Sophist-UK** [reply]
@smartavionics Mark

I quite understand your own position, but would you be willing to open up this repo to having a v5 branch that is community maintained?

This would retain the single point for an enhanced version of Cura and avoid multiple personal forks which end up fragmented and unmaintained?

**Reply by smartavionics** [reply]
> I quite understand your own position, but would you be willing to open up this repo to having a v5 branch that is community maintained?

Sorry, no.

Why can't someone just fork my repos and then they can do what they wish?

**Reply by Sophist-UK** [reply]
They can - that is an option. 

But one of the great things about your enhancements were that they were all in one place, with generation of executables for various platforms. 

If various people reimplement a subset of those same enhancements in their own repos that they want as an individual, we will end up with various repos with different enhancements, most of which will remain largely unknown, and (in all likelihood) these will over time become code-only without generation of executables because for a single user who is already running from source, maintaining the creation of executables for various platforms that they don't need is not going to be a priority.

Alternatively someone can create a Cura community organisation that will jointly maintain such a shared repo.

**Reply by PhilBaz** [reply]
Two thigs...

1) @smartavionics  I've deleted my old posts below as this thread has become something different. And it was probably the wrong place in the first place.

2) I get why everyone wants what they want. And if Mark wants to minimize hassle going forward, fair enough of course.

If there is a desire for a single point/fork incorporating Marks work into 5.0 perhaps someone can take on that role and mark can simply make reference to that in an appropriate post/thread/repository, wherever makes the most sense and is most easily found.  

Just a thought.

Ps. thanks again Smart!

**Reply by Sophist-UK** [reply]
If it will help, I am happy to do the admin of an organisation and create a community repo fork from Mark's, and to administer the security of who can e.g. approve PRs.

But I don't have time (and probably not the skills either) to code or review PRs etc.

**Reply by shyblower** [reply]
I've put my fork of the original Cura and CuraEngine repos on Github. I've called them SuperCura and SuperCuraEngine respectively. It's a pun on SuperSlicer, which is a fork of PrusaSlicer. I hope I'm allowed to do this. If Ultimaker doesn't like it I would of course rename it.
I've already pushed my port of the overhang fan speed and cooling fan lag features to the 5.2 branches and also have made them the default branches for now, so that people who miss these features in Cura 5 and are able to build the software by themselves, can already use it.

I still have to figure out how to correctly configure the available Github actions for OS package creation to make prebuilt executables available. During my initial attempt I had some success with creating the Windows installer, although I'm most proficient with Linux, but the Linux and MacOS build scripts need a PGP key for package signing (why for heavens sake the windows installer does not get signed?) and I haven't had the time yet, to look into that much further.
I also haven't modified the readme, to describe what's going on, yet.

Feel free to use it in whatever way you like ... as long as it's not illegal or immoral (I've just added this to not get myself in trouble in case someone uses it e.g. for obtaining world domination 🙃).

Thank you @smartavionics for letting me exploit your code, which made it quite easy for me to get these two features into Cura 5 as I had not too much to adapt ... some other feature ports might become more painful, though,  but I'm not in a hurry and am happy that I got the overhang fan speed override back, now that I've found out that even PrusaSlicer 😱  already supports it, tough PS won't let you configure the overhang angle threshold.

**Reply by Sophist-UK** [reply]
For the benefit of others the URLs are

https://github.com/shyblower/SuperCura

https://github.com/shyblower/SuperCuraEngine

Note: Obviously you can accept PRs on these repos and make other users who deserve it partial admins etc. and in the future if there is enough of a community you could create a SuperCura organisation and transfer the repos there for wider adoption.

**Reply by 5axes** [reply]
> people who miss these features in Cura 5 and are able to build the software by themselves, can already use it.

As they are a very few users Who can do that in the Cura community  the chance to have a real audience are very limited

**Reply by shyblower** [reply]
I'm working on it but probably not this week because I will be visiting the Formnext 2022 in Frankfurt/Germany and therefore be away for 4 days.

**Reply by Sophist-UK** [reply]
And I should also echo similar thanks myself to Mark for all his efforts. 👏👏👏👏

**Comment by smartavionics** [comment]
> Shwaa??? 4.20.4

> Arachne?

You talk in riddles, what are you asking?

**Comment by smartavionics** [comment]
> The inner wall seems to be a single wide extrusion. Arachne like. Have our dreams come true?

I don't think that's an inner wall, it looks to me like you have the fill gaps between walls feature enabled and my implementation of that works like a crude version of Arachne. That code predated Arachne but was never offered to UM because it was very much an experiment.
